<a href="https://colab.research.google.com/github/deepalikamble33/DataScience/blob/main/Assignment1_file2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

(Classifying Handwritten Digits from Tensorflow 2.x
• Try improving the model and code the program from scratch 
• Write up a summary explaining how your program works

In [2]:
import tensorflow as tf
from tensorflow import keras

Loading MNIST dataset

In [29]:
mnist=keras.datasets.mnist
(X_Train, Y_Train), (X_Test, Y_Test)= mnist.load_data()

In [30]:
print(X_Train.shape)
print(Y_Train.shape)
print(X_Test.shape)
print(Y_Test.shape)

(60000, 28, 28)
(60000,)
(10000, 28, 28)
(10000,)


In [31]:
X_Train=(X_Train.astype('float32'))
X_Test=(X_Test.astype('float32'))
X_Train= X_Train.reshape(60000, 784)
X_Test=X_Test.reshape(10000,784)

In [34]:
X_Train.shape

(60000, 784)

In [35]:
X_Train= X_Train/255
X_Test=X_Test/255

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]], dtype=float32)

In [36]:
Y_Train= tf.keras.utils.to_categorical(Y_Train)
Y_Test= tf.keras.utils.to_categorical(Y_Test)

In [65]:
import tensorflow as tf
from tensorflow import keras
from keras import Sequential
from keras.layers import Dense
from keras.layers import Dropout

In [66]:
model= keras.Sequential()
model.add(Dense(200, input_shape= (784,), name='dense_layer', activation='relu'))
model.add(Dropout(0.2))
model.add(Dense(10, name='dense_layer1', activation='softmax'))

In [67]:
# Compiling the model.
model.compile(optimizer='Adam', 
              loss='categorical_crossentropy',
              metrics=['accuracy'])

In [68]:
# Training the model. 
training = model.fit(X_Train, Y_Train, batch_size=32, epochs=40, validation_split=0.2) 

Epoch 1/40
1500/1500 [==============================] - 6s 4ms/step - loss: 0.2918 - accuracy: 0.9151 - val_loss: 0.1416 - val_accuracy: 0.9599
Epoch 2/40
1500/1500 [==============================] - 5s 3ms/step - loss: 0.1343 - accuracy: 0.9601 - val_loss: 0.1023 - val_accuracy: 0.9693
Epoch 3/40
1500/1500 [==============================] - 5s 4ms/step - loss: 0.0964 - accuracy: 0.9710 - val_loss: 0.0941 - val_accuracy: 0.9732
Epoch 4/40
1500/1500 [==============================] - 6s 4ms/step - loss: 0.0771 - accuracy: 0.9759 - val_loss: 0.0825 - val_accuracy: 0.9748
Epoch 5/40
1500/1500 [==============================] - 5s 4ms/step - loss: 0.0632 - accuracy: 0.9798 - val_loss: 0.0847 - val_accuracy: 0.9745
Epoch 6/40
1500/1500 [==============================] - 5s 3ms/step - loss: 0.0543 - accuracy: 0.9823 - val_loss: 0.0857 - val_accuracy: 0.9760
Epoch 7/40
1500/1500 [==============================] - 5s 3ms/step - loss: 0.0450 - accuracy: 0.9859 - val_loss: 0.0809 - val_accuracy:

In [69]:
#evaluate the model
test_loss, test_acc = model.evaluate(X_Test, Y_Test)
print('Test accuracy:', test_acc)

313/313 [==============================] - 1s 3ms/step - loss: 0.1234 - accuracy: 0.9803
Test accuracy: 0.9803000092506409


In [70]:
model.summary()

Model: "sequential_5"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_layer (Dense)         (None, 200)               157000    
                                                                 
 dropout (Dropout)           (None, 200)               0         
                                                                 
 dense_layer1 (Dense)        (None, 10)                2010      
                                                                 
Total params: 159,010
Trainable params: 159,010
Non-trainable params: 0
_________________________________________________________________


Adding hidden layer increases accuracy to 97 from 92 as it learns patterns in data to greater extend. By adding more neuron to hidden layer, further help in improving the accuracy to 98% but execution time increased. Epoch reduced from 70 to 40 so to reduce execution time without much compromising the accuracy. Batch size reduced from 64 to 32 which help to understand the pattern and slight increase in accuracy. With dropout layer accuracy improved and helps to avoid over-fitting of data. 